test/e2e/util/util.go

```diff
@@ -283,3 +290,144 @@ func EnsureNodeCountMatchesNodePoolReplicas(t *testing.T, ctx context.Context, h
+		}
+
+		// Get thanos-querier endpoint
+		promRoute := &routev1.Route{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "thanos-querier",
+				Namespace: "openshift-monitoring",
+			},
+		}
+		if err := client.Get(ctx, crclient.ObjectKeyFromObject(promRoute), promRoute); err != nil {
+			t.Skip("unable to get prometheus route, skipping")
+		}
+		promEndpoint := fmt.Sprintf("https://%s", promRoute.Status.Ingress[0].Host)
+
+		// Create prometheus client
+		cfg := promconfig.HTTPClientConfig{
+			Authorization: &promconfig.Authorization{
+				Type:        "Bearer",
+				Credentials: promconfig.Secret(token),
+			},
+			TLSConfig: promconfig.TLSConfig{
+				InsecureSkipVerify: true,
+			},
+		}
+		rt, err := promconfig.NewRoundTripperFromConfig(cfg, "e2e-budget-checker")
+		if err != nil {
+			t.Fatalf("failed to get create round tripper: %v", err)
+		}
+		client, err := promapi.NewClient(promapi.Config{
+			Address:      promEndpoint,
+			RoundTripper: rt,
+		})
+		if err != nil {
+			t.Fatalf("failed to get create prometheus client: %v", err)
+		}
+		v1api := promv1.NewAPI(client)
+
+		// Compare metrics against budgets
+		namespace := manifests.HostedControlPlaneNamespace(hostedCluster.Namespace, hostedCluster.Name).Name
+		budgets := []struct {
+			name   string
+			query  string
+			budget float64
+		}{
+			{
+				name:   "control-plane-operator read",
+				query:  fmt.Sprintf("sum(hypershift:controlplane:component_api_requests_total{method=\"GET\", namespace=~\"%s\"}) by (namespace)", namespace),
+				budget: 1000, // upgrade tests make this high, typically under 200 for non-upgrade tests
+			},
+			{
+				name:   "control-plane-operator mutate",
+				query:  fmt.Sprintf("sum(hypershift:controlplane:component_api_requests_total{method!=\"GET\", namespace=~\"%s\"}) by (namespace)", namespace),
+				budget: 750, // upgrade tests make this high, typically under 200 for non-upgrade tests
+			},
+			{
+				name:   "control-plane-operator no 404 deletes",
+				query:  fmt.Sprintf("sum(hypershift:controlplane:component_api_requests_total{method=\"DELETE\", code=\"404\", namespace=~\"%s\"}) by (namespace)", namespace),
+				budget: 0,
+			},
+			// hypershift-operator budget can not be per HC so metric will be significantly under budget for all but the last test(s) to complete on a particular test cluster
+			// These budgets will also need to scale up with additional tests that create HostedClusters
+			{
+				name:   "hypershift-operator read",
+				query:  "sum(hypershift:operator:component_api_requests_total{method=\"GET\"})",
+				budget: 600,
+			},
+			{
+				name:   "hypershift-operator mutate",
+				query:  "sum(hypershift:operator:component_api_requests_total{method!=\"GET\"})",
+				budget: 1500,
+			},
+			{
+				name:   "hypershift-operator no 404 deletes",
+				query:  "sum(hypershift:operator:component_api_requests_total{method=\"DELETE\", code=\"404\"})",
+				budget: 0,
+			},
+		}
+
+		for _, budget := range budgets {
```

Comment:
Maybe make these subtests `(t.Run(budged.name, func(t *testing.T){testcode})` so this gets more legible in the error case?